keras_rs/src/layers/modeling/dot_interaction.py

```diff
@@ -64,52 +103,50 @@ def call(self, inputs: list[types.Tensor]) -> types.Tensor:
         # Check if all feature tensors have the same shape and are of rank 2.
         shape = ops.shape(inputs[0])
         for idx, tensor in enumerate(inputs):
-            if ops.shape(tensor) != shape:
+            other_shape = ops.shape(tensor)
+
+            if len(shape) != 2:
+                raise ValueError(
+                    "All feature tensors inside `inputs` should have rank 2. "
+                    f"Received rank {len(shape)} at index {idx}."
+                )
+
+            if not check_shape(shape, other_shape):
                 raise ValueError(
                     "All feature tensors in `inputs` should have the same "
                     f"shape. Found at least one conflict: shape = {shape} at "
-                    f"index 0 and shape = {ops.shape(tensor)} at index {idx}"
+                    f"index 0 and shape = {ops.shape(tensor)} at index {idx}."
                 )
 
-        if len(shape) != 2:
-            raise ValueError(
-                "All feature tensors inside `inputs` should have rank 2. "
-                f"Received rank {len(shape)}."
-            )
-
         # `(batch_size, num_features, feature_dim)`
         features = ops.stack(inputs, axis=1)
 
-        batch_size, _, _ = ops.shape(features)
+        batch_size, num_features, _ = ops.shape(features)
 
         # Compute the dot product to get feature interactions. The shape here is
         # `(batch_size, num_features, num_features)`.
         pairwise_interaction_matrix = ops.matmul(
             features, ops.transpose(features, axes=(0, 2, 1))
         )
 
-        # if `self.self_interaction` is `True`, keep the main diagonal.
-        k = -1
-        if self.self_interaction:
-            k = 0
-
-        tril_mask = ops.tril(
-            ops.ones_like(pairwise_interaction_matrix, dtype=bool),
-            k=k,
-        )
-
         # Set the upper triangle entries to 0, if `self.skip_gather` is True.
         # Else, "pick" only the lower triangle entries.
         if self.skip_gather:
+            tril_mask = self._generate_tril_mask(pairwise_interaction_matrix)
+
             activations = ops.multiply(
                 pairwise_interaction_matrix,
                 ops.cast(tril_mask, dtype=pairwise_interaction_matrix.dtype),
             )
+            # Rank-2 tensor.
+            activations = ops.reshape(activations, (batch_size, -1))
```

Comment:
I've been avoiding using `-1` in reshape when it is possible to calculate the explicit value. The reason is that on Tensorflow, when you have a dynamic batch size (which is often the case with a tf dataset), `(batch_size, -1)` will make the second dimension dynamic too even though it can be computed statically. Then, the output will be fully dynamic with a static shape of `(None, None)`.

Instead, put the explicit value (I believe it's simply `num_features * num_features`). Note that this also works if `num_features` happens to be dynamic.

If we want to verify that what I'm saying is actually true :) here is how you test it:
https://github.com/keras-team/keras/blob/master/keras/src/layers/preprocessing/stft_spectrogram_test.py#L294

Reply:
Interesting. `None * None` will work?

Reply:
Oh, what I meant is that if it's dynamic, `ops.shape` will return a tensor of rank 0 for that dimension, multiplying tensors like that is allowed, and then `ops.reshape` accepts dimensions that are dynamic and specified via a tensor.